mrrc/storage.py

```diff
@@ -0,0 +1,235 @@
+"""
+Copyright (c) 2021 Red Hat, Inc
+All rights reserved.
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+import logging
+import os
+from typing import Callable, Dict, List
+
+from boto3 import session
+from botocore.errorfactory import ClientError
+
+from mrrc.utils.files import read_sha1
+
+logger = logging.getLogger(__name__)
+
+PRODUCT_META_KEY = "rh-products"
+CHECKSUM_META_KEY = "checksum"
+
+
+class S3Client(object):
+    """The S3Client is a wrapper of the original boto3 s3 client, which will provide
+       some convenient methods to be used in the mrrc uploader.
+    """
+
+    def __init__(self, extra_conf=None) -> None:
+        s3_session = session.Session()
+        self.client = s3_session.resource(
+            's3',
+            config=extra_conf,
+        )
+
+    def upload_files(self, file_paths: List[str], bucket_name: str, product: str, root="/"):
+        """ Upload a list of files to s3 bucket. * Use the cut down file path as s3 key. The cut
+        down way is move root from the file path if it starts with root. Example: if file_path is
+        /tmp/maven-repo/org/apache/.... and root is /tmp/maven-repo Then the key will be
+        org/apache/.....
+
+            * The product will be added as the extra metadata with key "rh-products". For
+            example, if the product for a file is "apache-commons", the metadata of that file
+            will contain "rh-products":"apache-commons"
+
+            * For existed files, the upload will not override them, as the metadata of
+            "rh-products" will be updated to add the new product. For example, if an exited file
+            with new product "commons-lang3" is uploaded based on existed metadata
+            "apache-commons", the file will not be overridden, but the metadata will be changed to
+            "rh-products": "apache-commons,commons-lang3"
+
+            * Every file has sha1 checksum in "checksum" metadata. When uploading existed files,
+            if the checksum does not match the existed one, will not upload it and report error.
+            Note that if file name match
+        """
+        bucket = self.__get_bucket(bucket_name)
+
+        def path_upload_handler(full_file_path: str, path: str):
+            if not os.path.isfile(full_file_path):
+                logger.warning('Warning: file %s does not exist during uploading. Product: %s',
```

Comment:
There is a \#TODO comment here before, please do not remove it as we need it as a reminder to fix the missing logic later

Reply:
I removed it because it was causing a linting issue. Can we add it as a requirement for one of the ongoing tasks so that it's completed before release? My thinking is that if we add a TODO in code then more likely than not we'll end up just leaving it there.